### ISSUE TITLE
Add printer friendly view

### ISSUE THREAD

**Issue by wurstchristoph** [owncloud/mail#717]
To allow printing of mails either a special print style sheet or a dedicated button for showing only the mail content is needed

**Comment by sndrsk** [comment]
I would also really appreciate a printer-friendly view. I also print to PDFs a lot, it would be great if I could download the e-mail as a PDF or save it to my cloud as a PDF.

**Comment by jancborchardt** [comment]
Yeah, good call. Tagging it as starter issue. Basically what has to be done is to hide everything except the subject, to/cc and the message body.

**Comment by colmoneill** [comment]
Could be fun to tackle this. What would be best then? Ctrl + P? a button to toggle print styles? Do we want this as a visible option of the unique email interface, or just something to appear for those of you who do print emails and do the keyboard shortcut?

**Comment by wurstchristoph** [comment]
@colmoneill I don't think any shortcut is needed, you can simply create a CSS file that only shows elements that should be visible in the print view.
http://www.w3.org/TR/CSS2/media.html

**Comment by colmoneill** [comment]
Cool.
I'm going to add a specific stylesheet for print, as mail.css is already nearing 700 lines.
https://doc.owncloud.org/server/8.1/developer_manual/app/css.html#css
Sound ok?

**Comment by colmoneill** [comment]
... if I can find the darn file through which I include new stylesheets...
Any ideas?

**Comment by wurstchristoph** [comment]
@jancborchardt what do you think? ^

@colmoneill it's added via the controller https://github.com/owncloud/mail/blob/master/lib/controller/pagecontroller.php#L66 ;-)

**Comment by jancborchardt** [comment]
@colmoneill yup, print.css is a good call. :)

**Comment by colmoneill** [comment]
@wurstchristoph @jancborchardt 

adding it into the pagecontroller.php does work
```
<link rel="stylesheet" href="/owncloud-core/apps/mail/css/mail.css?v=8d29749f5ecf4eff70cbe3882c393f55" media="screen">
<link rel="stylesheet" href="/owncloud-core/apps/mail/css/mobile.css?v=8d29749f5ecf4eff70cbe3882c393f55" media="screen">
<link rel="stylesheet" href="/owncloud-core/apps/mail/css/print.css?v=8d29749f5ecf4eff70cbe3882c393f55" media="screen">
```
but it tags the stylesheet as media, which is far from ideal.

I'm looking for a solution.

**Comment by wurstchristoph** [comment]
@colmoneill hm, seems like ``addStyle`` is depricated anyway (https://github.com/owncloud/core/blob/master/lib/private/appframework/core/api.php#L77-L88)

@jancborchardt do you know more about that? Should we move it to the template?

**Comment by jancborchardt** [comment]
Hm, @DeepDiver1975 probably knows best.

**Comment by DeepDiver1975** [comment]
Use style () in the template. Awesome guys - I'm honestly impressed. Thanks

**Comment by wurstchristoph** [comment]
@colmoneill @jancborchardt https://github.com/owncloud/core/pull/16857 should resolve the media issue

**Comment by colmoneill** [comment]
Ey cool. I admit I kind of let this go while researching the media issue.

Now that this is in place, I'll try and make some time to make a suggestion.

**Comment by wurstchristoph** [comment]
@colmoneill great, thanks!

**Comment by Tramour** [comment]
Any news on this issue? I'd appreciate it very much to be able to convert emails to pdf and store them directly in the cloud.

**Comment by jancborchardt** [comment]
@Tramour the support for print stylesheets will arrive earliest in 8.2: https://github.com/owncloud/core/pull/16857 (the version we’ll release the next days is 8.1)

But then – @colmoneill do you want to take a stab at a print stylesheet? As I said above:

> Basically what has to be done is to hide everything except the current email: subject, to/cc and the message body.

**Comment by Tramour** [comment]
Okay. Thanks for the answer.

**Comment by colmoneill** [comment]
Hey all,

Indeed, I have started this, but ran up against an issue that I think must be linked to the content-overflow nature of the mail app columns.

![screenshot from 2015-07-24 14 47 30](https://cloud.githubusercontent.com/assets/2122318/8874633/ee231e90-3212-11e5-9d19-13f7bd92158a.png)

Hiding content from the rest of the app is easy, but I can't work out how to get text to flow from one page to an other. In regular print css, this is dealt with with the `page` property, but this cant be called here, without rewriting a part of the markup; something I think we can all agree would be preferable.

Meanwhile, css regions were a glimmer of hope, but got shot down and now only exist in the form of (heavy) polyfils. This could be a solution but I'd rather ask other from the front end team if they could take a look at why they think the content is not flowing. It works fine for short emails of course :)

I'm in an not-up-to-date fork https://github.com/colmoneill/mail/tree/printStylesheet in branch printStylesheet.

Tip: continuing my love/hate relationship with chrome, it is the better browser to test print styles, it renders a print preview straight after a `Ctrl`+ `p` and the device mode inspector has a print stylesheet toggle which is clearly helpful.

**Comment by jancborchardt** [comment]
@colmoneill what’s the exact issue here though? Looks fine except some details (line on top, subject not showing completely).

**Comment by colmoneill** [comment]
Sorry: the issue is that if the quantity of content exceeds the size of the page, it just gets hidden, it does not flow over to a new page. Multi page pdf is super easy for browsers, and I can't work out why oC mail would have such an issue. I believe it is linked to the way content is placed absolutely, with overflows allowed, but I tried working it out and couldn't.

CSS print is kind of a black box in that sense, it's not easy to get proper debugging tools, because the majority of the formatting (page size, margins, header, footer, date time stamps) are just one click away from being overwritten by the browser.

**Comment by colmoneill** [comment]
Edit: scratch that. I finally found what the issue was. All the positionning being absolute, and then with internal div content flow, the browser kept a property that made it think that it should not extend further than 100% of the viewport. In the case of print, that was an A4 page.

I can continue on now :) I just had to let it be for a __long__ while. Ill see about putting together enough for a PR this weekend.

**Comment by jancborchardt** [comment]
@colmoneill cool stuff  :)

**Comment by ChristophWurst** [comment]
@colmoneill any updates? :)

**Comment by colmoneill** [comment]
Hey all, sorry for being so absent, specifically on the things I put my hand up for. In fact, I even fell behind in core versions of my own oC, but did an update yesterday, super impressed with 8.2 and I'm now making some time to finish this print stylesheet for mail over the winter vacation.

Cheers!

**Comment by jancborchardt** [comment]
Yeeeha @colmoneill, looking forward! Hope to meet you at FOSDEM – if so, consider submitting a proposal for the [Open Source Design room](http://opensourcedesign.net/events/design/hack/meeting/2016/01/30/opensourcedesign-fosdem.html)!

**Comment by ChristophWurst** [comment]
@colmoneill ping ;-)

**Comment by ChristophWurst** [comment]
Merged: https://github.com/owncloud/core/pull/16857
